src/mongo_odm/query_builder.hpp

```diff
@@ -24,94 +25,462 @@
+    const int &_remainder;
+};
+
+/**
+ * Represents a query that performs a text search with the $text operator.
+ * TODO check values of `language` against spec?
+ */
+class TextSearchExpr {
+   public:
+    /**
+     * Creates a text search expression.
+     * These parameters correspond to the parameters for the $text operator in MongoDB.
+     * @param  search               A string of terms use to query the text index.
+     * @param  language             The language of the text index.
+     * @param  case_sensitive      A boolean flag to specify case-sensitive search. Optional,
+     * false
+     *                             by default.
+     * @param  diacritic_sensitive A boolean flag to specify case-sensitive search. Optional,
+     * false
+     *                             by default.
+     */
+    constexpr TextSearchExpr(const char *search, const char *language, bool case_sensitive = false,
+                             bool diacritic_sensitive = false)
+        : _search(search),
+          _language(language),
+          _case_sensitive(case_sensitive),
+          _diacritic_sensitive(diacritic_sensitive) {
+    }
+
+    /**
+     * Creates a text search expression, with the default `language` setting of the text index.
+     * These parameters correspond to the parameters for the $text operator in MongoDB.
+     * @param  search               A string of terms use to query the text index.
+     * @param  case_sensitive      A boolean flag to specify case-sensitive search. Optional,
+     * false
+     *                             by default.
+     * @param  diacritic_sensitive A boolean flag to specify case-sensitive search. Optional,
+     * false
+     *                             by default.
+     */
+    constexpr TextSearchExpr(const char *search, bool case_sensitive = false,
+                             bool diacritic_sensitive = false)
+        : _search(search),
+          _case_sensitive(case_sensitive),
+          _diacritic_sensitive(diacritic_sensitive) {
+    }
+
+    /**
+     * Appends this expression to a BSON core builder,
+     * as a key-value pair of the form " key: { $mod: [ divisor, remainder ] } ".
+     * @param builder a BSON core builder
+     * @param wrap      Whether to wrap the BSON inside a document.
+     */
+    void append_to_bson(bsoncxx::builder::core &builder, bool wrap = false) const {
+        if (wrap) {
+            builder.open_document();
+        }
+        builder.key_view("$text");
         builder.open_document();
-        builder.key_view(_selector_type);
-        builder.append(_field);
+        builder.key_view("$search");
+        builder.append(_search);
+        if (_language) {
+            builder.key_view("$language");
+            builder.append(_language.value());
+        }
+        builder.key_view("$caseSensitive");
+        builder.append(_case_sensitive);
+        builder.key_view("$diacriticSensitive");
+        builder.append(_diacritic_sensitive);
```

Comment:
Since $caseSensitive and $diacriticSensitive are false by default, it would probably save some bandwidth over the wire if you only included them if they're true.